lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -324,16 +327,46 @@ def _on_update_status(self, event: UpdateStatusEvent):
         # handle when/if certificates are expired
         self._check_certs_expiration(event)
 
-    def _on_get_admin_secrets_action(self, event: ActionEvent):
+    def _on_set_password_action(self, event: ActionEvent):
+        """Set new admin password from user input or generate if not passed."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+
+        user_name = event.params.get("username")
+        if event.params.get("username") != "admin":
+            event.fail("Only the 'admin' username is allowed for this action.")
+            return
+
+        password = event.params.get("password") or generate_password()
+        try:
+            self._put_admin_user(password)
+            password = self.secrets.get(Scope.APP, "admin_password")
+            event.set_results({f"{user_name}-password": password})
+        except OpenSearchError as e:
+            event.fail(f"Failed changing the password: {e}")
+
+    def _on_get_password_action(self, event: ActionEvent):
         """Return the password and cert chain for the admin user of the cluster."""
-        password = self.secrets.get(Scope.APP, "admin_password")
+        if event.params.get("username") != "admin":
+            event.fail("Only the 'admin' username is allowed for this action.")
+            return
 
-        chain = ""
-        admin_secrets = self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val)
-        if admin_secrets and admin_secrets.get("chain"):
-            chain = "\n".join(admin_secrets["chain"][::-1])
+        if not self._is_tls_fully_configured():
+            event.fail("admin user or TLS certificates not configured yet.")
+            return
 
-        event.set_results({"password": password if password else "", "chain": chain})
+        admin_password = self.secrets.get(Scope.APP, "admin_password")
```

Comment:
**suggestion** maybe f"{username}_password" to be parametric, if we plan to support multiple usernames

Reply:
My understanding is that we do not plan to having more users in OpenSearch (outside the client relation) - and that we'll only have the `admin` user, I'll confirm this and report back